src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/ProjectTypeCapabilitiesProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks.Dataflow;
+
+using ProjectCapabilitiesProjectValue = Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.IProjectCapabilitiesSnapshot>;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Web
+{
+    /// <summary>
+    ///     Produces capabilities from project type GUIDs.
+    /// </summary>
+    [Export(ExportContractNames.Scopes.ConfiguredProject, typeof(IProjectCapabilitiesProvider))]
+    [AppliesTo(ProjectCapability.DotNet)]
+    internal class ProjectTypeCapabilitiesProvider : ChainedProjectValueDataSourceBase<IProjectCapabilitiesSnapshot>, IProjectCapabilitiesProvider
+    {
+        private readonly ProjectTypeGuidsDataSource _dataSource;
+
+        [ImportingConstructor]
+        protected ProjectTypeCapabilitiesProvider(ConfiguredProject configuredProject, ProjectTypeGuidsDataSource dataSource)
+            : base(configuredProject)
+        {
+            _dataSource = dataSource;
+        }
+
+        public ProjectCapabilitiesProjectValue? Current { get; private set; }
+
+        protected override IDisposable LinkExternalInput(ITargetBlock<ProjectCapabilitiesProjectValue> targetBlock)
+        {
+            JoinUpstreamDataSources(_dataSource);
+
+            DisposableValue<ISourceBlock<ProjectCapabilitiesProjectValue>> block = _dataSource.SourceBlock.Transform(
+                snapshot => Current = snapshot.Derive(s => CreateCapabilitiesSnapshot(s)));
+
+            block.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+
+            return block;
+        }
+
+        private IProjectCapabilitiesSnapshot CreateCapabilitiesSnapshot(IImmutableList<Guid> projectTypes)
+        {
+            ImmutableHashSet<string> capabilities = Empty.CapabilitiesSet;
+
+            // TODO: Maybe pull this from a specific item type, or metadata on a capability?
+            if (projectTypes.Contains(ProjectType.LegacyWebApplicationProjectGuid))
+            {
+                capabilities = capabilities.Add("AspNet");
```

Comment:
Just a general question. where in the load sequence are capability providers loaded? Do all components which are associated with these capabililties need to be DyanamicLoad components because the capability can come and go? Or are they considered stable? I would really prefer the latter.

Reply:
These capabilities are not fixed, they are dynamic. This is very deliberate - this will mean that a project behaves identically between switching to/from a branch with `<ProjectTypeGuids/>` property and loading a project for the first time with a `<ProjectTypeGuids/>`. Otherwise, they will behave differently. If components/feature have too much problems handling this downstream, then we can cause a reload of the project based on `ProjectTypeGuids` changing, but I'd like that as a last resort.